### PR TITLE
[FSDP2] Added HSDP grad acc tests and some minor changes

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125484
* __->__ #125479
* #125431

This adds HSDP to the existing gradient accumulation tests and includes some minor changes to simplify things a tiny bit.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/testing/_internal/common_fsdp.py [hunk 1]
```diff
@@ -907,6 +907,18 @@ def patch_reduce_scatter(new_reduce_scatter_tensor: Callable):
         dist.reduce_scatter_tensor = orig_reduce_scatter
 
 
+@contextlib.contextmanager
+def patch_all_reduce(new_all_reduce: Callable):
```
Comment (wanchaol) [inline torch/testing/_internal/common_fsdp.py:911]: hopefully after https://github.com/pytorch/pytorch/pull/125475 landed lots of these would be simplified!

**Review by wanchaol — APPROVED** [review]
